client/wsclient.go

```diff
@@ -245,19 +250,33 @@ func (c *wsClient) runOneCycle(ctx context.Context) {
 		c.common.PackagesStateProvider,
 		c.common.Capabilities,
 	)
-	r.ReceiverLoop(ctx)
-
-	// Stop the background processors.
-	procCancel()
-
-	// If we exited receiverLoop it means there is a connection error, we cannot
-	// read messages anymore. We need to start over.
+	r.Start(ctx)
+
+	select {
+	case <-c.sender.IsStopped():
+		// sender will send close message to initiate the close handshake
+		if err := c.sender.Err(); err != nil && err != websocket.ErrCloseSent {
```

Comment:
Is this `err != websocket.ErrCloseSent` to handle the case when the server initiates the close?

Reply:
Yes. The sender and receiver run concurrently. There is a possibility that the server and client initiate the closing handshake at the same time. In this case, the close message may be sent by CloseHandler, which is called in the receiver.

Reply:
Can we change `sender.Err()` to return nil if the error was `websocket.ErrCloseSent`? Let's consider it an internal implementation detail of the sender and avoid leaking it here. That we only need to check `err` for nil here. Sender seeing a `websocket.ErrCloseSent` response is a success, we don't need to consider it an error.